doc: Drop description of LogError messages as fatal

ryanofsky opened the pull request:
Remove documentation that says `LogError` should be used for "severe problems that require the node (or a subsystem) to shut down entirely" because:

- It is not how `LogError` and `Level::Error` are used currently. Of 129 current uses only 58 cases are fatal according to [#30348](https://github.com/bitcoin/bitcoin/issues/30348)

- "[T]here's not much benefit in a log that says "hey this error is about to cause the node to stop" -- you already get that information by seeing "Shutdown: in progress..." immediately following." according to [#30347 (comment)](https://github.com/bitcoin/bitcoin/pull/30347#discussion_r1659096893)

Fixes #30348

ryanofsky committed:
doc: Drop description of LogError messages as fatal

Remove documentation that says LogError should be used for "severe problems
that require the node (or a subsystem) to shut down entirely" because:

- This is not how `LogError` and `Level::Error` are used currently. Of 129
  current uses only 58 cases are fatal according to
  https://github.com/bitcoin/bitcoin/issues/30348

- "[T]here's not much benefit in a log that says "hey this error is about to
  cause the node to stop" -- you already get that information by seeing
  "Shutdown: in progress..." immediately following." according to
  https://github.com/bitcoin/bitcoin/pull/30347#discussion_r1659096893

ajtowns commented:
I think both entries should continue to include examples.

I think removing "severe" from the LogWarning description doesn't make sense -- anything the admin might have to address is a severe (potential) problem, isn't it?

Having LogError mean "the node admin **will** need to address" vs LogWarning mean "the node admin **may** need to address" seems to argue towards retaining the "LogError is fatal" interpretation to me -- fatal errors are ones the admin does need address; anything else they can ignore until it becomes fatal.

A potential advantage of having LogError (or LogFatal) reserved for fatal errors is that you could grep through the codebase for LogError to see all the potentially fatal errors, and perhaps work on finding ways to make them less fatal. Not convinced that's very beneficial though.

I don't think there's much other benefit to splitting potential problems up from actual problems otherwise if the admin still needs to attend to them either way, though; so if we're already making LogError just mean actual errors; is there any reason not to merge them both into a single LogAlert() or similar?

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30361#issuecomment-2197781492

Thanks, added back examples and "severe".

I think probably just looking at AbortNode callers would be as effective as as searching for fatal error messages, if looking for places to make error handling more robust.

I do like the alert level idea and posted an implementation in #30364.

Updated 5175cbcbb00d6212cf548de428f43cb0c2a1be4b -> b7aae361b273f2f439d3b278214b7e37908c8cb0 ([`pr/nofat.1`](https://github.com/ryanofsky/bitcoin/commits/pr/nofat.1) -> [`pr/nofat.2`](https://github.com/ryanofsky/bitcoin/commits/pr/nofat.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofat.1..pr/nofat.2)) with clarifications and examples.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30361#issuecomment-2231748835

I think I might not have enough information to understand the weak nack, because I don't actually see a reason in there to reject a documentation fix that is only dropping advice to interpret `Error` as "fatal" and `Warning` as "nonfatal".

The advice is not followed by current code, but more importantly, I don't think anybody (even the original author) currently thinks this advice is something we should try to implement anymore. If something in this PR makes the documentation worse, it'd be helpful to know where the problems are. Otherwise this seems like a clean fix for the issue described in #30348, and a way to close that issue and move on to other fixes and improvements.

I do understand what you are saying about the `error()` removal in #29236 making the issue described in #30348 worse because it increased the amount of Error spam. But it's not like this PR is changing documentation to say that Error spam is ok. It is clearly not. It is just trying to remove a separate point of confusion.

ajtowns commented:
> I don't think anybody (even the original author) currently thinks this advice is something we should try to implement anymore.

Presuming that if it's me who's he-who-must-not-be-named here rather than referring back to #24464 etc, what I think is that if we're going to have distinct log levels for "error" and "warning", I think there should be a clear documented difference between what they are, so that PR authors, reviewers and users can know what to expect. If there's places in our code that don't meet that expectation, then that's something that should be fixed, though it doesn't need to be treated as some huge priority. On the other hand, if there's no distinction between the levels that's worth making, we shouldn't have two levels. So I think either the status quo or #30364 is better than this PR. It's not really worth arguing about, but it's frustrating having people put words in my mouth that aren't mine.

l0rinc commented on doc/developer-notes.md:
```diff
@@ -742,14 +743,13 @@ logging messages. They should be used as follows:
-- `LogError(fmt, params...)` should be used in place of `LogInfo` for
-  severe problems that require the node (or a subsystem) to shut down
-  entirely (e.g., insufficient storage space).
+- `LogError(fmt, params...)` should be used in place of `LogInfo` for severe
+  errors the node admin will need to address (e.g., failure to write data).
```
nit: *`instead of`* seems [slightly more common](https://www.italki.com/en/post/question-107941) than *`in place of`* + `that` to streamline readability + `needs to address` to imply urgency:
```suggestion
- `LogError(fmt, params...)` should be used instead of `LogInfo` for severe errors
  that the node admin needs to address (e.g., failure to write data).
```

l0rinc commented on doc/developer-notes.md:
```diff
@@ -742,14 +743,13 @@ logging messages. They should be used as follows:
+  should address (e.g. system time appears to be wrong, unknown soft fork
+  appears to have activated).
 
 - `LogTrace(BCLog::CATEGORY, fmt, params...)` should be used in place of
```
nit: the logger order is weird, why debug -> info -> error -> warning -> trace?
Especially after announcing `LogInfo`, `LogDebug`, `LogTrace`, `LogWarning` and `LogError` in the intro.

sedited reviewed (approved):
<!--meta-tag:bot-skip-->

ACK 5e2eb5bde2f08c8a55430d3cd4e4454f1180ebd4

maflcko reviewed (commented):
I am still not sure about this. I understand that the distinction is not clear in the code, but I think the docs can serve as a guide to what the "final" code should look like. I don't think we should "weaken" the docs, just because the current code isn't ideal.
The fist section of the dev notes also explain this process:

```md
## Coding Style (General)

Various coding styles have been used during the history of the codebase,
and the result is not very consistent. However, we're now trying to converge to
a single style, which is specified below. When writing patches, favor the new
style over attempting to mimic the surrounding style, except for move-only
commits.

```


Also, this is sold as a fix for https://github.com/bitcoin/bitcoin/issues/30348, but most of the points in there seem to be outdated/stale?

sedited commented:
> For example, if the point is that (let's say) an index write io error should not lead to a crash, then that seems reasonable. However, just logging and continuing normally (hiding the error log in the scrollback) doesn't seem ideal either. Whether that log is a warning or error may not matter much at all? Possibly it is better to disconnect the index, so that no stale results are silently returned?

Another example are a few of the validation failures. For example we don't shutdown in all cases where a block read from disk subsequently fails `CheckBlock()`. This should require intervention, since it should be highly unusual for a node to let a bad block in in the first place. A warning might be more appropriate there too, but I think a log level providing distinction between severe and minor issues makes sense for things like these.

ajtowns commented:
Concept NACK. 

Most of the issues here -- that is, where not-very-important errors are logged via `LogError` -- were introduced via scripted-diff in #29236 without putting much thought into whether the old `error()` and the then-new `LogError` were a good match. This change just retroactively makes that PR's acks be in line with the docs.

But logging not-very-important errors as `LogError` (or even `LogWarning`) is not a desirable thing for our codebase; the entire value of tagging some log messages as warning/error is that it's important for them to be acted upon.

I don't think this set of issues and PRs (#30347, #30348, #30361, #30364) has achieved anything except to serve as an ongoing distraction from improving those log messages.

sedited commented:
Retracting my A-C-K here. Seems like we were closer to realizing the log level guidelines than I thought. Given the two standing N-A-C-Ks, I think this should be closed.

sedited commented:
@ryanofsky what is your intent here? As this stands, it has 3 NACKs.
